File: blog_app_flask/app/models/blogpostmodel.py
from app.extensions import db
from datetime import datetime
from app.models.usermodel import User

class PostModel(db.Model):
    __tablename__ = 'posts'
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(100), nullable=False)
    content = db.Column(db.Text, nullable=False)
    image = db.Column(db.String(255))
    author_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
    date_created = db.Column(db.DateTime, default=datetime.now)

    author = db.relationship('User', backref=db.backref('posts', lazy=True))

    def comment_count(self):
        return len(self.comments)

    def __repr__(self):
        return f'<Post {self.title}>'


class Comments(db.Model):
    __tablename__ = 'comments'
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
    post_id = db.Column(db.Integer, db.ForeignKey('posts.id', ondelete='CASCADE'), nullable=False)
    content = db.Column(db.String(200), nullable=False)
    date_created = db.Column(db.DateTime, default=datetime.now)

    user = db.relationship('User', backref=db.backref('user_comments', lazy=True, cascade="all, delete"))
    post = db.relationship('PostModel', backref=db.backref('comments', lazy='dynamic', cascade="all, delete"))

    def __repr__(self):
        return f'<Comment {self.content}>'

File: blog_app_mern/api/models/BlogpostModel.js
import mongoose, { Schema } from 'mongoose';

const postSchema = new Schema({
    title: { type: String, required: true, maxlength: 100 },
    content: { type: String, required: true },
    image: { type: String, maxlength: 255 },
    author: { type: Schema.Types.ObjectId, ref: 'users', required: true },
    dateCreated: { type: Date, default: Date.now },
    comments: [{ type: Schema.Types.ObjectId, ref: 'Comment' }]
});

postSchema.virtual('commentCount').get(function () {
    return this.comments.length;
});

const commentSchema = new Schema({
    user: { type: Schema.Types.ObjectId, ref: 'users', required: true },
    post: { type: Schema.Types.ObjectId, ref: 'Post', required: true },
    content: { type: String, required: true, maxlength: 200 },
    dateCreated: { type: Date, default: Date.now }
});

postSchema.pre('remove', function(next) {
    this.model('Comment').deleteMany({ post: this._id }, next);
});

commentSchema.pre('remove', function(next) {
    this.model('Post').updateOne({ _id: this.post }, { $pull: { comments: this._id } }, next);
});

const Post = mongoose.model('Post', postSchema);
const Comment = mongoose.model('Comment', commentSchema);

export { Post, Comment };

